### PR TITLE
Rewrite PasswordHash.Parse to work with ReadOnlySpans

### PR DESCRIPTION
**Pull request by Bond-009**

### CODE REVIEW

File: MediaBrowser.Common/Cryptography/PasswordHash.cs [hunk 1]
```diff
@@ -59,58 +70,109 @@ public PasswordHash(string id, byte[] hash, byte[] salt, Dictionary<string, stri
...
+                        parametersSpan = parametersSpan[(index + 1)..];
                     }
 
-                    string[] fields = paramset.Split('=');
-                    if (fields.Length != 2)
+                    int splitIndex = parameter.IndexOf('=');
+                    if (splitIndex == -1 || splitIndex == 0 || splitIndex == parameter.Length - 1)
                     {
-                        throw new InvalidDataException($"Malformed parameter in password hash string {paramset}");
+                        throw new FormatException($"Malformed parameter in password hash string");
```
Comment (crobibero) [inline MediaBrowser.Common/Cryptography/PasswordHash.cs:126]:
```suggestion
                        throw new FormatException("Malformed parameter in password hash string");
```

File: MediaBrowser.Common/Cryptography/PasswordHash.cs [hunk 2]
```diff
@@ -59,58 +70,109 @@ public PasswordHash(string id, byte[] hash, byte[] salt, Dictionary<string, stri
...
+                    return new PasswordHash(id.ToString(), Array.Empty<byte>(), Array.Empty<byte>(), parameters!);
                 }
+
+                hashString = hashString[(nextSegment + 1)..];
+                nextSegment = hashString.IndexOf('$');
+            }
+
+            if (nextSegment == 0)
+            {
+                throw new FormatException($"Hash string contains an empty segment");
```
Comment (crobibero) [inline MediaBrowser.Common/Cryptography/PasswordHash.cs:146]:
```suggestion
                throw new FormatException("Hash string contains an empty segment");
```

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.Common/Cryptography/PasswordHash.cs [hunk 3]
```diff
@@ -30,6 +31,16 @@ public PasswordHash(string id, byte[] hash, byte[] salt)
...
         {
+            if (id == null)
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
+
+            if (id.Length == 0)
+            {
+                throw new ArgumentException("String can't be empty", nameof(id));
+            }
```
Comment (barronpm) [inline MediaBrowser.Common/Cryptography/PasswordHash.cs:42]: Why not `string.IsNullOrEmpty`?

Comment (Bond-009) [reply MediaBrowser.Common/Cryptography/PasswordHash.cs:42]: We already checked `null`

Comment (barronpm) [reply MediaBrowser.Common/Cryptography/PasswordHash.cs:42]: Right, but it could just be a call to `IsNullOrEmpty` instead of both of these checks

Comment (Bond-009) [reply MediaBrowser.Common/Cryptography/PasswordHash.cs:42]: Different exceptions

Comment (barronpm) [reply MediaBrowser.Common/Cryptography/PasswordHash.cs:42]: Is there a benefit to doing that? the string is invalid either way, and the rest of the code is written the way I described. If this is the way we want to go, that's fine, but we should be consistent about it.

Comment (Bond-009) [reply MediaBrowser.Common/Cryptography/PasswordHash.cs:42]: More expressive, consistent with the .Net BCL?

**Review by barronpm — APPROVED** [review]

**Pull request merged**